File: cama-tablet/android/app/src/main/java/com/cama/tablet/ble/TabletDeviceInfo.kt
package com.cama.tablet.ble

import android.bluetooth.BluetoothAdapter
import android.content.Context
import android.provider.Settings
import org.json.JSONObject

data class TabletDeviceInfo(
    val deviceId: String,
    val deviceName: String,
    val serviceUuid: String,
    val dataCharUuid: String,
) {
    fun toQrPayloadJson(): String {
        val json = JSONObject()
        json.put("v", BleConstants.QR_VERSION)
        json.put("app", BleConstants.QR_APP_ID)
        json.put("deviceId", deviceId)
        json.put("deviceName", deviceName)
        json.put("serviceUuid", serviceUuid)
        json.put("dataCharUuid", dataCharUuid)
        return json.toString()
    }

    companion object {
        fun from(context: Context): TabletDeviceInfo {
            val androidId = Settings.Secure.getString(
                context.contentResolver,
                Settings.Secure.ANDROID_ID,
            ) ?: "unknown"

            val adapter = BluetoothAdapter.getDefaultAdapter()
            val name = adapter?.name?.takeIf { it.isNotBlank() }
                ?: "CAMA-Tablet-${androidId.takeLast(4).uppercase()}"

            return TabletDeviceInfo(
                deviceId = androidId,
                deviceName = name,
                serviceUuid = BleConstants.SERVICE_UUID.toString(),
                dataCharUuid = BleConstants.HEALTH_DATA_CHAR_UUID.toString(),
            )
        }
    }
}

File: cama-tablet/ios/CamaTablet/BLE/TabletDeviceInfo.swift
import Foundation
import UIKit

struct TabletDeviceInfo {
    let deviceId: String
    let deviceName: String
    let serviceUuid: String
    let dataCharUuid: String

    func toQrPayloadJson() -> String {
        let payload: [String: Any] = [
            "v": BleConstants.qrVersion,
            "app": BleConstants.qrAppId,
            "deviceId": deviceId,
            "deviceName": deviceName,
            "serviceUuid": serviceUuid,
            "dataCharUuid": dataCharUuid,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return json
    }

    func toDictionary() -> [String: Any] {
        [
            "deviceId": deviceId,
            "deviceName": deviceName,
            "serviceUuid": serviceUuid,
            "dataCharUuid": dataCharUuid,
        ]
    }

    static func current() -> TabletDeviceInfo {
        let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? "unknown"
        let suffix = String(deviceId.suffix(4)).uppercased()
        let name = UIDevice.current.name.trimmingCharacters(in: .whitespacesAndNewlines)
        let deviceName = name.isEmpty ? "CAMA-Tablet-\(suffix)" : name

        return TabletDeviceInfo(
            deviceId: deviceId,
            deviceName: deviceName,
            serviceUuid: BleConstants.serviceUUID.uuidString,
            dataCharUuid: BleConstants.healthDataCharUUID.uuidString
        )
    }
}
